### ISSUE TITLE
chrome 124: weird results

### ISSUE THREAD

**Issue by krausest** [krausest/js-framework-benchmark#1655]
**Management Summary:**

- Chrome 124 results don't look right
- The test runner yields results that differ from manual testing for create 1k rows (see [comment](https://github.com/krausest/js-framework-benchmark/issues/1655#issuecomment-2075611512) below)
- Frameworks that use RAF might be reported as faster than frameworks without RAF (see imba below). I didn't look into that issue in detail. The create 1k rows issue is currently more important to me.

**Details:**

A first run for the keyed implementations show results that don't look good.
The fastest frameworks score now 1.05. In the history this has been a bad sign that something might be odd with the measurement.
The fastest implementation for clear row was imba:
![Screenshot 2024-04-20 at 19 48 18](https://github.com/krausest/js-framework-benchmark/assets/3374055/87929119-729c-4986-a249-69d8ddd1b5ee)
Imba uses RAF to schedule the rendering. In the screenshot above we'd compute 14.95 msecs for the duration (from click event to the end of the first commit event). Please note that there's a separate hit test and commit event later in the trace that doesn't count for our computation at least at the moment.

Vanillajs has only one commit event and looks more straight forward:
![Screenshot 2024-04-20 at 19 47 55](https://github.com/krausest/js-framework-benchmark/assets/3374055/29f14ce7-6da3-4d73-812a-50adbf5838ab)
So here we'd take 17.38 msecs for vanillajs.

For chrome 123 we measured 15.3 for vanillajs and 16.7 for imba.
With chrome 124 the results look different:
<img width="300" src="https://github.com/krausest/js-framework-benchmark/assets/3374055/13ca5a8b-04a5-4d1e-8930-a3bafea96afe">
The boxplot looked like that for chrome 123:
<img width="300" src="https://github.com/krausest/js-framework-benchmark/assets/3374055/dbce7d5d-041b-4c00-a6e3-c27251f88b54">

**Comment by krausest** [comment]
If I add a RAF to vanillajs I get results pretty close to imba and I get a second commit event.
![Screenshot 2024-04-20 at 20 20 59](https://github.com/krausest/js-framework-benchmark/assets/3374055/fb1acd04-b8eb-4a8f-b673-569d07e46935)
Maybe it was a wrong decision to count to the first commit and ignore the second commit in this case.

**Comment by krausest** [comment]
Ran the benchmark for the keyed frameworks again last night. At least results are very consistent ☹️ 
Here's a screenshot:
![Screenshot 2024-04-22 at 20 41 21](https://github.com/krausest/js-framework-benchmark/assets/3374055/b76d827d-8f7f-48af-8ab8-35285ea800d3)
or [here](https://krausest.github.io/js-framework-benchmark/2024/chrome124_possibly_incorrect.html) if you want to look at the table.
It'll take some time for further analysis.

**Comment by krausest** [comment]
Here's a look at create rows:
A rather normal vanillajs trace:
![Screenshot 2024-04-22 at 21 13 07](https://github.com/krausest/js-framework-benchmark/assets/3374055/3c63b6e3-877f-448f-a6bd-d81d26297ede)
The fastest ivi trace:
![Screenshot 2024-04-22 at 21 12 56](https://github.com/krausest/js-framework-benchmark/assets/3374055/45d6548e-d547-4de9-b2d6-63fae74b0abf)
<img width=363 src="https://github.com/krausest/js-framework-benchmark/assets/3374055/9b59fc20-bd6a-4f49-9d96-b4a1293a7ec2">
This wasn't the case with chrome 123:
<img width=363 src="https://github.com/krausest/js-framework-benchmark/assets/3374055/604513ff-24fb-493a-8c71-7c75ecb4fd98"/>

With manual testing I haven't managed to get faster results for ivi, it looks more like chrome 123.

**Comment by antonmak1** [comment]
@krausest Maybe this has something to do with the HTML and DOM changes in version 124? Specifically, "Document render-blocking" or "setHTMLUnsafe and parseHTMLUnsafe" or something like that?

**Comment by krausest** [comment]
I took a closer look at the create 1k rows issue - but I have no good news:
I ran a manual test for create 1k rows with 8 runs for ivi, solid, doohtml and vanillajs. This is what a boxplot looks like for this manual test. I'd consider that to be the ground truth:

<img width=500 src="https://github.com/krausest/js-framework-benchmark/assets/3374055/e90adbfc-6371-4e4e-b8f4-34751685a956">

Please note that there's a suspicious outlier for vanillajs with 34.29 msecs. I kept a screenshot of that trace.
There's a quite clear ordering between the other three: doohtml < ivi < solid

What's interesting is that the chart above is remarkably similar to the puppeteer chrome 123 results:
<img width=500 src="https://github.com/krausest/js-framework-benchmark/assets/3374055/27eb2923-6971-4f1b-8fa4-0ae4071a2869">

However the results for all testdrivers are disappointing. I performed three runs for each of them:

1. Puppeteer: "clearly ivi is fastest"

![puppeteer124](https://github.com/krausest/js-framework-benchmark/assets/3374055/7b65c49c-dd4e-4b96-9e3a-fe842dd1ef6d)
At least results are repeatable though not similar to the manual results.

2. Playwright: "solid is fastest"

![playwright124](https://github.com/krausest/js-framework-benchmark/assets/3374055/28d8429d-4b8e-4924-b25f-c9aa7edb8b6c)

3. Webdrivercdp: "where is my mind?"
![webdrivercdp124](https://github.com/krausest/js-framework-benchmark/assets/3374055/a887fd24-b1f6-4157-8e74-1735fff62c8d)
The first run fits to "doo < ivi < solidjs" though vanillajs is closer to the outlier from the manual testing above, but run two and three just look random.

Currently I'm out of ideas. I don't see how I could publish chrome 124 results soon.

**Comment by antonmak1** [comment]
@krausest Perhaps, if the old code works somehow wrong, then first of all, it needs to be tested together with the new code, which needs to be written based on version 124. It is clear that some kind of nonsense is coming out, that all the results now start from 1.05 and those frameworks and libraries that were plus or minus in one place are now 10 positions ahead, then 10 positions behind - this is nonsense. How can 7+ releases be adequate, but then everything breaks with the new one? This means that there is definitely a new error somewhere in the code that was not an error before or was not considered one.

**Comment by mksunny1** [comment]
I will add my mind here. See if I can come up with something. I am not too familiar with this yet.

**Comment by krausest** [comment]
There's not much code that was updated: https://github.com/krausest/js-framework-benchmark/commit/30c247a897cfe147a94ff82edafd8d066f4ffb36
I also tried to update all the dependencies, but it doesn't change anything so I rolled them back. So I really think the factor that makes the difference is chrome 124.

I tried if windows reports better results, but it doesn't look like it does.

**Comment by krausest** [comment]
Just when I finished my plan what I'd do after quitting this project I tried one last thing 😄 
What if I add a sleep call of a second before each interaction?
Turns out that this results in the following charts for three independent runs for puppeteer:
![slow_runs](https://github.com/krausest/js-framework-benchmark/assets/3374055/d8a36ca4-ca5c-4382-a1e7-050f75d01172)
Those charts look much closer to manual testing (and of course as old as I'm feeling that second between clicks seems realistic).

Now I have to find out:

- How does the sleep impact the chart of the other testdrivers?
- Where do we need the sleep calls? (hopefully not at many places)
- What's the minimum duration for the sleep?

**Comment by krausest** [comment]
I don't see that the delay helps the other test drivers.
playwright with a delay of 1 second:
![playwright_slow](https://github.com/krausest/js-framework-benchmark/assets/3374055/0a47bfb1-2549-4d2f-81c4-2ae1ae097269)
Looks different from the chart without delays, but not close to the manual testing result.

Nor does webdrivercdp:

![webdrivercdp_slow](https://github.com/krausest/js-framework-benchmark/assets/3374055/07f92316-85ce-4b5b-9799-8e3c849982e6)

**Comment by krausest** [comment]
2. Chrome 124 is bizarre. I tried first adding sleeps between all interactions and then trying to filter out where they are actually needed.

I got that chart with puppeteer:
![Screenshot 2024-04-27 at 14 41 33](https://github.com/krausest/js-framework-benchmark/assets/3374055/9f520f39-653d-486b-80ae-882e37346041)

Look how bad those numbers are: > 50 msecs. That's ridiculous.
Here's a trace for one such bad run:
![Screenshot 2024-04-27 at 14 45 47](https://github.com/krausest/js-framework-benchmark/assets/3374055/e5b32b81-50a9-4c5a-b53e-55df3b3c14b2)
And here's the trace file:
[doohtml-keyed_01_run1k_5.json](https://github.com/krausest/js-framework-benchmark/files/15138212/doohtml-keyed_01_run1k_5.json)
The trace looks right, the computation of the duration is OK, there's no RAF, no GC, nothing suspicious. But scripting 5 ms and rendering 48 ms is just incredibly bad.

**Comment by krausest** [comment]
3. The disaster is complete
![chrome_124_disaster](https://github.com/krausest/js-framework-benchmark/assets/3374055/0dad3a4a-aa67-4851-a8d6-907d56cb5827)
On the right are chrome 123 results. The left four charts are chrome 124 with varying sleep calls after every click during initBenchmark and before and after forceGC. Chrome 123 used no sleeps and thus should correspond to no sleep.
I've painted a baseline since the axis is different for the charts.
Sleep with 1 second creates the chart most similar to chrome 123, but is slower that the others and chrome 123.
Basically I'd say that the fastest results should be considered most correct (I mean Vmax means here t_min, no matter how it's achieved).

Not sure how to get out of here. Currently I see no chance that chrome 124 allows me to measure with any confidence.

**Comment by titoBouzout** [comment]
May I suggest to try chrome 126 (chrome dev channel), to see if the problem persist? I'm not sure if worthy but you can also try beta and canary channels. Maybe they already fixed something? What about experiments, last time I read about experiments you cannot turn them off.

**Comment by krausest** [comment]
This comment serves just as a summary and will be linked from the chrome issue report. It contains no new information to the above.

With chrome 124 I got strange results for the benchmark that can be be seen in the chart below:
![chrome_124_disaster](https://github.com/krausest/js-framework-benchmark/assets/3374055/0dad3a4a-aa67-4851-a8d6-907d56cb5827)
The results for chrome 123 on the right should be seen as a baseline. The chart "no sleep" is using the same code as chrome 123 and should be identical, but it is far off!
I tried if delaying the actions from the benchmark driver helps and indeed adding a sleep of one second gives an order similar to chrome 123 but at a different speed. Varying the sleep duration makes the ranking of the frameworks arbitrary: ivi can be fastest, slowest and third. There's a chart above that shows that chrome 123 are stable for those sleep durations.

The duration for the benchmark is measured via traces from the click to the paint commit event:
![Screenshot 2024-04-29 at 21 00 55](https://github.com/krausest/js-framework-benchmark/assets/3374055/341663a2-979b-4f45-8cd3-92a75dae7174)
For chrome 124 with 500 msecs delay we get something like 38.81 for one benchmark run
Without sleep it takes only 34.78 msecs.
![Screenshot 2024-04-29 at 21 01 35](https://github.com/krausest/js-framework-benchmark/assets/3374055/c5a67cb9-d617-4bfe-bc01-8275b7f4a08a)
Please note that the difference comes from rendering duration, though rendering shouldn't behave different when some sleeps delays are added before the click event.

**Comment by krausest** [comment]
Chrome issue report is here: https://issues.chromium.org/issues/337900449

**Comment by krausest** [comment]
@titoBouzout I tried that before, but didn't keep the result. Anyways I repeated the run. It looks like that without sleep:
<img width="400" src="https://github.com/krausest/js-framework-benchmark/assets/3374055/965f92b7-adcc-454a-bfdd-ce36bb3f8360">
Just as bad as chrome 124
@localvoid Any idea why ivi is most impacted?

**Comment by localvoid** [comment]
@krausest I am not sure about other libraries, but I think that the only difference (DOM operations) between ivi and vanillajs is that when table is cleared, it removes rows by replacing `<tbody>` DOM node with a new one instead of removing rows with `textContent=""`

**Comment by krausest** [comment]
I'm really stuck because I have no confidence in the chrome 124+ results, but I have a proposal to resolve that:
If we achieve to create a new vanillajs implementation that performs create 1,000 rows, replace rows and create 10,000 rows faster (or as fast) as ivi in the no delay scenario above I'll gain back confidence. If we fail I claim that chrome reports incorrect values.
@syduki @trueadm @localvoid or anyone else: Any chance submitting a new vanillajs version?
(One more rule: Just cloning ivi and using that as a lib wouldn't count 😄 )
I opened an issue for that: https://github.com/krausest/js-framework-benchmark/issues/1661

**Comment by krausest** [comment]
Thanks to you (especially @robbiespeed) we found something: The GC didn't work properly for ivi. Tweaking the GC calls made results much more as expected.

I uploaded new results for keyed implementations here: https://krausest.github.io/js-framework-benchmark/2024/table_chrome_124_preview2.html

Looks much more reasonable. Event the remove row benchmarks looks better, imba is back to normal.

There's still one decision open:
Should we compute RAF frameworks like bobril with 22.64 msecs (from click to end of 2nd commit)
![Screenshot 2024-05-08 at 21 40 21](https://github.com/krausest/js-framework-benchmark/assets/3374055/ca425095-8ceb-41a4-bc1e-0a23357a4f0e)
or as 16.04 msecs (from click to end of first commit)
![Screenshot 2024-05-08 at 21 40 47](https://github.com/krausest/js-framework-benchmark/assets/3374055/279e8bc6-9d25-45e8-ba9e-baa82dc54574)

So far we take the 2nd approach, but if the 2nd commit happens for all RAF frameworks I think switching to alternative 1 seems more logical to me. 
Seems like there's no second commit for frameworks that don't use RAF.

I'll check both assumption and report back.

**Comment by krausest** [comment]
One last check. It's really crazy. Forcing GC should help most frameworks, since it prevents GC during execution of the framework. Except that it doesn't help ivi...

Duration in msecs
| Framework    | no GC | 7x window.gc() | Full GC |
| -------- | ------- | ------- | ------- |
| vanillajs  |  41.3   | 38 |  38 |
| ivi  |  37.6   | 35.2 | 39.2 |

Full GC = window.gc({type:'major',execution:'sync',flavor:'last-resort'}), which has the same effect as a loop of HeapProfiler.collectGarbage and window.gc()

Memory after create 1k
| Framework    | no GC | 7x window.gc() | Full GC |
| -------- | ------- | ------- | ------- |
| vanillajs  |  1.78  | 1.81 |  1.77 |
| ivi  |  2.09   | 2.09 | 2.07 |

Anyways I'm closing this issue now, since the chrome 124 anomalies were removed with the new Full GC.
